### PR TITLE
MOABB: Redesign DataIO for better efficiency and flexibility

### PR DESCRIPTION
**Pull request by Drew-Wagner**
Redesigns the dataio pipeline to make use of speechbrain's DynamicItemDataset, with a philosophy of natively supporting lazy loading (for large datasets).

**NOTE:** this PR does not integrate the new data pipeline with the training scripts.

### CODE REVIEW

**Comment by bruAristimunha** [comment]
FYI: @agramfort, the draft of the lazy dataloader using the MNE-BIDs

**Comment by Drew-Wagner** [comment]
@bruAristimunha

Currently, the epochs are lazily loaded as `mne.io.Raw` objects using the `.crop(...)` method.
```python
        @sb.utils.data_pipeline.takes("raw", "onset")
        @sb.utils.data_pipeline.provides("epoch")
        def _load_epoch(raw, onset):
            onset_time = onset / raw.info["sfreq"]
            return raw.copy().crop(
                tmin=onset_time + tmin,
                tmax=onset_time + tmax if tmax is not None else None,
                verbose=False,
            )
```

The advantage of this is that it allows us to more easily access all the MNE methods and avoid loading the epoch to memory. The disadvantage however, is that it seems that the `.crop(...)` method has significantly more overhead than simple slicing. In my opinion, it then makes sense to instead of cropping return the epoch as a numpy array:
```python
        @sb.utils.data_pipeline.takes("raw", "onset")
        @sb.utils.data_pipeline.provides("epoch", "times")
        def _load_epoch(raw: mne.io.RawArray, onset: int):
            onset_time = onset / raw.info["sfreq"]
            epoch, times = raw[
                :,
                raw.time_as_index(onset_time + tmin) : (
                    raw.time_as_index(onset_time + tmax)
                    if tmax is not None
                    else -1
                ),
            ]
            times -= onset_time + tmin
            yield epoch
            yield times
```

What are your thoughts?

File: benchmarks/MOABB/dataio/datasets.py [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+                bids_path, extra_params=dict(preload=False), verbose=False
+            )
+            stim_channels = mne.utils._get_stim_channel(
+                None, raw.info, raise_error=False
+            )
+            if len(stim_channels) > 0:
+                # returns empty array if none found
+                events = mne.find_events(raw, shortest_event=0, verbose=False)
+                event_id = {}
+            else:
```
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/datasets.py:382]: Do we really need this?

File: benchmarks/MOABB/dataio/datasets.py [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+
+        Returns
+        -------
+        dict[str, dict]
+            Returns a mapping from unique sample ID to sample metadata
+        """
+        json_data: dict[str, dict] = {}
+
+        matched_paths: Iterable[BIDSPath] = bids_path.update(
+            suffix=cls.suffix
```
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/datasets.py:286]: The suffix here, is it what we are looking?

Comment (bruAristimunha) [reply benchmarks/MOABB/dataio/datasets.py:286]: If we are only creating the metainformation to the eeg file (e.g., .edf), we are not mapping the channel information from the bid folds (which could be useful in the future) or the events file that is created (which could be helpful to now). I understand the tradeoff here about the events, because in many moabb datasets, the files are small, and perhaps it is more practical to just recreate the events based on the channel.

File: benchmarks/MOABB/dataio/datasets.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+        ).match(ignore_json=True)
```
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/datasets.py:287]:
```suggestion
        matched_paths: Iterable[BIDSPath] = mne_bids.find_matching_paths(
            root=bids_path.root,
        )
```
this will return everything

File: benchmarks/MOABB/dataio/datasets.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+        for path in matched_paths:
+            uid = path.fpath.name
+            json_data[uid] = path.entities
+            json_data[uid]["fpath"] = str(path.fpath)
+        return json_data
```
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/datasets.py:292]: I liked this, I think I will integrate this on moabb

File: benchmarks/MOABB/dataio/datasets.py (hunk 1 above)
Comment (bruAristimunha) [reply benchmarks/MOABB/dataio/datasets.py:382]: When we create the bid folder with moabb, we don't have the stimulus channels anymore.

https://github.com/NeuroTechX/moabb/blob/e2080202b62946a123b5870da1c3646f4f3a20b9/moabb/datasets/bids_interface.py#L337

File: benchmarks/MOABB/dataio/datasets.py [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+            # TODO: How to handle the case where multiple values map to the same key?
+            event_id = {v: k for k, v in event_id.items()}
+
+            for onset, _, event in events:
+                label = event_id.get(event, int(event))
+                event_sample = dict(**sample, label=label, onset=int(onset))
+                event_uid = f"{uid}/{label}/{onset}"
+                json_data[event_uid] = event_sample
+
+        return json_data
```
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/datasets.py]:
```suggestion

```
We have saved the information in the folder path. I don't think we need to recreate it.

File: benchmarks/MOABB/dataio/splitters.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    @property
+    def targets(self) -> Sequence[tuple[TargetT]]:
+        return tuple(combinations(super().targets, self.leave_k_out))
+
+    def _get_test_data_ids(self, target):
+        test_data_ids = tuple(
+            chain.from_iterable(self.folds[t] for t in target)
+        )
+
+        return test_data_ids
```
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/splitters.py]: I understand that we have already talked about this and that with this, we generate the permutation of all possible train and test data; however, I feel that it is more intuitive to use a sklearn object to do the splitter. At least, it is much more common.

If others are okay with this splitter in this combinatorial way, and subsequent filtering, I'm okay, otherwise, let's use sklearn. pinging @mravanelli, @vmcru, @ddavidebb, @engmubarak48, @Drew-Wagner

File: benchmarks/MOABB/dataio/datasets.py
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/datasets.py:1]: - unit tests; (80% or more)
- doc string with example;

File: benchmarks/MOABB/dataio/datasets.py (hunk 2 above)
Comment (Drew-Wagner) [reply benchmarks/MOABB/dataio/datasets.py:286]: When we call `read_raw_bids` it looks in the channels.tsv and events.tsv files. I believe it also looks at the coordinate system, etc.

My understanding is that the only thing that is thrown away are the non-eeg channels (like meg channels for example)

File: benchmarks/MOABB/dataio/datasets.py (hunk 1 above)
Comment (Drew-Wagner) [reply benchmarks/MOABB/dataio/datasets.py:382]: I kept it this way since we may want to support datasets which do not come from MOABB as well

File: benchmarks/MOABB/dataio/datasets.py (hunk 5 above)
Comment (Drew-Wagner) [reply benchmarks/MOABB/dataio/datasets.py]: I don't understand what you are proposing here...

File: benchmarks/MOABB/dataio/datasets.py [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+------
+Drew Wagner, 2025
+Bruno Aristimunha, 2025
+"""
+from __future__ import annotations
+
+import json
+from functools import cache
+from pathlib import Path
+from typing import Any, Hashable, Iterable, Optional, Self, TypedDict
```
Comment (vmcru) [inline benchmarks/MOABB/dataio/datasets.py]: Self for typing only works for python 3.11 and not for 3.10

File: benchmarks/MOABB/dataio/datasets.py [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+
+    @classmethod
+    def from_moabb(
+        cls,
+        dataset: BaseMOABBDataset,
+        json_path: str | Path,
+        subjects=None,
+        save_path: Optional[str] = None,
+        **cls_kwargs,
+    ) -> Self:
```
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/datasets.py]:
```suggestion
    ):
```

File: benchmarks/MOABB/dataio/datasets.py [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+                    pass
+
+    @classmethod
+    def from_bids(
+        cls,
+        bids_path: BIDSPath | Path | str,
+        json_path: Path | str,
+        subjects=None,
+        **cls_kwargs,
+    ) -> Self:
```
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/datasets.py]:
```suggestion
    ):
```

File: benchmarks/MOABB/dataio/datasets.py [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+
+    @classmethod
+    def from_moabb(
+        cls,
+        dataset: BaseMOABBDataset,
+        json_path: str | Path,
+        subjects=None,
+        save_path: str | None = None,
+        **cls_kwargs,
+    ) -> Self:
```
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/datasets.py]:
```suggestion
    ):
```

File: benchmarks/MOABB/dataio/datasets.py (hunk 7 above)
Comment (bruAristimunha) [inline benchmarks/MOABB/dataio/datasets.py]:
```suggestion
from typing import Any, Hashable, Iterable, Optional, TypedDict
```
